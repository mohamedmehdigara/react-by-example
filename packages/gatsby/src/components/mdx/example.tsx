import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
