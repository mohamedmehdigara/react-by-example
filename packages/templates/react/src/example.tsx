import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
