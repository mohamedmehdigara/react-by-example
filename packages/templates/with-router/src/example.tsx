import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
