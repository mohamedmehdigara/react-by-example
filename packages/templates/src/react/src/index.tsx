import React from 'react'
import ReactDOM from 'react-dom'
import example from './example'

ReactDOM.render(<>{example}</>, document.getElementById('root'))
