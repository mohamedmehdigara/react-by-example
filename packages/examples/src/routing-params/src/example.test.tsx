import { cleanup, render, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { FC } from 'react'
import { MemoryRouter } from 'react-router'
import { Params } from './example'

describe('Params', () => {
  let component: RenderResult

  interface TestProps {
    location: string
  }
  const Test: FC<TestProps> = ({ location }) => {
    return (
      <MemoryRouter initialEntries={[location]}>
        <Params />
      </MemoryRouter>
    )
  }

  afterEach(() => {
    cleanup()
  })

  describe('when home', () => {
    beforeEach(() => {
      component = render(<Test location="/" />)
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    describe('when clicking a user', () => {
      beforeEach(() => {
        userEvent.click(component.getByText('Alice'))
      })

      it('should snapshot', () => {
        expect(component.container).toMatchSnapshot()
      })

      describe('when clicking an article', () => {
        beforeEach(() => {
          userEvent.click(component.getByText("Alice's second article"))
        })

        it('should snapshot', () => {
          expect(component.container).toMatchSnapshot()
        })

        describe('when clicking username', () => {
          beforeEach(() => {
            userEvent.click(component.getByText('by Alice'))
          })

          it('should go back to article listing', () => {
            expect(component.container).toHaveTextContent(/Posts from Alice/)
          })
        })
      })
    })
  })

  describe('when no articles', () => {
    beforeEach(() => {
      component = render(<Test location="/user/carol" />)
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })
  })

  describe('when user not found', () => {
    beforeEach(() => {
      component = render(<Test location="/user/dave" />)
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })
  })

  describe('when article not found', () => {
    beforeEach(() => {
      component = render(<Test location="/article/99" />)
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })
  })

  describe('when 404', () => {
    beforeEach(() => {
      component = render(<Test location="/404" />)
    })

    it('should redirect to users list', () => {
      expect(component.getByRole('heading')).toHaveTextContent('Users')
    })
  })
})
