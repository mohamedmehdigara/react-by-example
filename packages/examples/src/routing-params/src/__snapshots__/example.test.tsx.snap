// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Params when article not found should snapshot 1`] = `
<div>
  <div>
    <h1>
      Cannot find article
    </h1>
  </div>
</div>
`;

exports[`Params when home should snapshot 1`] = `
<div>
  <div>
    <h1>
      Users
    </h1>
    <ul>
      <li>
        <a
          href="/user/alice"
        >
          Alice
        </a>
      </li>
      <li>
        <a
          href="/user/bob"
        >
          Bob
        </a>
      </li>
      <li>
        <a
          href="/user/carol"
        >
          Carol
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Params when home when clicking a user should snapshot 1`] = `
<div>
  <div>
    <div>
      <h1>
        Posts from 
        Alice
      </h1>
      <ul>
        <li>
          <a
            href="/article/1"
          >
            Alice's first article
          </a>
        </li>
        <li>
          <a
            href="/article/3"
          >
            Alice's second article
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Params when home when clicking a user when clicking an article should snapshot 1`] = `
<div>
  <div>
    <div>
      <h1>
        Alice's second article
      </h1>
      <p>
        Pellentesque lorem ligula, molestie nec tincidunt vitae, convallis eget mauris.
Mauris vitae nibh eleifend, pellentesque dolor sed, sollicitudin ipsum. Proin
luctus ipsum mauris, nec dignissim nibh suscipit et. Fusce luctus, justo sed
consectetur volutpat, arcu lectus molestie arcu, vel pulvinar urna eros eu sem.
Aliquam at est egestas, interdum odio eu, rutrum nulla. Quisque ac nisi eros.
Aliquam venenatis pulvinar ex, sed sagittis nibh viverra non. Lorem ipsum dolor
sit amet, consectetur adipiscing elit. Donec ipsum sapien, viverra ac bibendum
sit amet, varius vitae tortor.
      </p>
      <a
        href="/user/alice"
      >
        <i>
          by 
          Alice
        </i>
      </a>
    </div>
  </div>
</div>
`;

exports[`Params when no articles should snapshot 1`] = `
<div>
  <div>
    <div>
      <h1>
        Posts from 
        Carol
      </h1>
      <p>
        Carol
         has no articles
      </p>
    </div>
  </div>
</div>
`;

exports[`Params when user not found should snapshot 1`] = `
<div>
  <div>
    <h1>
      Cannot find user with id 
      dave
    </h1>
  </div>
</div>
`;
